### PR TITLE
Added SourceLinesOfCode metric

### PR DESCRIPTION
**Pull request by AlbertoMonteiro**

### CODE REVIEW

**Comment by AlbertoMonteiro** [comment]
Jacob, I fixed calculation.

I am using spaces instead of tabs, but there is a trick in GitHub to ignore whitespace in diff view, if you add "w=1" query string in URL, GitHub will ignore whitespaces and you will have much more cleaner diff.

Open this: https://github.com/jjrdk/ArchiMetrics/pull/9/files
And this: https://github.com/jjrdk/ArchiMetrics/pull/9/files?w=1

Using the last link you can see changes in project files.

Am I missing something?

File: ArchiMetrics.CodeReview.Rules.Tests/ArchiMetrics.CodeReview.Rules.Tests.csproj [hunk 1]
```diff
@@ -1,194 +1,197 @@
...
+      <Link>Hunspellx64.dll</Link>
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
+    <Content Include="..\packages\NHunspell.1.2.5359.26126\content\Hunspellx86.dll">
+      <Link>Hunspellx86.dll</Link>
+      <CopyToOutputDirectory>Always</CopyToOutputDirectory>
+    </Content>
+  </ItemGroup>
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Comment (jjrdk) [inline ArchiMetrics.CodeReview.Rules.Tests/ArchiMetrics.CodeReview.Rules.Tests.csproj]: What is this service?

File: ArchiMetrics.Analysis/Metrics/TypeMetricsCalculator.cs [hunk 2]
```diff
@@ -41,7 +41,8 @@ public TypeMetricsCalculator(SemanticModel semanticModel, Solution solution)
 			var source = CalculateClassCoupling(type, memberMetrics);
 			var depthOfInheritance = CalculateDepthOfInheritance(type);
 			var cyclomaticComplexity = memberMetrics.Sum(x => x.CyclomaticComplexity);
-			var linesOfCode = memberMetrics.Sum(x => x.LinesOfCode);
+			var linesOfCode = memberMetrics.Sum(x => x.LinesOfCode);
+            var sourceLinesOfCode = memberMetrics.Sum(x => x.SourceLinesOfCode);
```
Comment (jjrdk) [inline ArchiMetrics.Analysis/Metrics/TypeMetricsCalculator.cs]: If you do want to calculate SLOC, then this is incorrect as you are not capturing the class declaration lines.

File: ArchiMetrics.Analysis/Metrics/NamespaceMetricsCalculator.cs [hunk 3]
```diff
@@ -29,7 +29,8 @@ public NamespaceMetricsCalculator(SemanticModel semanticModel)
 		public INamespaceMetric CalculateFrom(NamespaceDeclarationSyntaxInfo namespaceNode, IEnumerable<ITypeMetric> metrics)
 		{
 			var typeMetrics = metrics.AsArray();
-			var linesOfCode = typeMetrics.Sum(x => x.LinesOfCode);
+			var linesOfCode = typeMetrics.Sum(x => x.LinesOfCode);
+            var sourceLinesOfCode = typeMetrics.Sum(x => x.SourceLinesOfCode);
```
Comment (jjrdk) [inline ArchiMetrics.Analysis/Metrics/NamespaceMetricsCalculator.cs]: Similar to error in type metrics calculation, this calculation does not include namespace declaration lines. Furthermore, using statements may happen outside of the namespace declaration, where do they get counted? In project metrics?

**Comment by jjrdk** [comment]
Nice trick with the whitespace, but I would still appreciate it if you could follow the coding style already in place.

**Comment by AlbertoMonteiro** [comment]
Now I guess it is fine.
Btw, about namespace node I am ignoring using directives.

**Comment by jjrdk** [comment]
You SLOC calculation seems arbitrary. When I look into what is written about SLOC calculation, I find that the definition is typically non-blank and non-comment lines in source files. Your calculation takes everything.

I would suggest making it more fine grained. This would also make it possible to do some metrics around comment and white space density.

What about converting the type to a SourceCodeStatistics type which could then include these more detailed metrics and which can then be extended in the future with other statistics, ex keyword density, etc.

**Comment by AlbertoMonteiro** [comment]
Well Jacob, I added a new commit that subtract comments of SourceLinesOfCode in MemberMetric.
I guess that we could improve this library in small steps. 
At moment these modifications that I made, fits fine for my project https://github.com/AlbertoMonteiro/CodeMetricsExtractor.
How about create a issue about SourceCodeStatistics and wait for more contributions. I would love still working on this wonderful library, but at moment I cant spend much time on some metrics that aren't going to fit on my need.

**Comment by AlbertoMonteiro** [comment]
@jjrdk none answer?

**Comment by jjrdk** [comment]
It looks like a nice project, but I would like to keep the API of my project in line with its purpose. In this question, adding a somewhat arbitrary SLOC value does not add value.

As I said, if the feature were developed to provide actual source code statistics, then I could see some value. For the moment, I will leave this open until the feature fits well.

**Pull request closed without merging**